url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/environment/Camera.java

```diff
@@ -0,0 +1,882 @@
+
+        P = new Matrix4x4();
+        aspect = viewportXSize / viewportYSize; 
+        if ( projectionMode == PROJECTION_MODE_PERSPECTIVE ) {
+            upDistance = nearPlaneDistance * Math.tan(Math.toRadians(fov/2));
+            downDistance = -upDistance;
+            leftDistance = aspect * downDistance;
+            rightDistance = aspect * upDistance;
+            P.frustumProjection(leftDistance, rightDistance,
+                                downDistance, upDistance,
+                                nearPlaneDistance, farPlaneDistance);
+            msg = msg + "  - Projection matrix:" + P;
+          }
+          else if ( projectionMode == PROJECTION_MODE_ORTHOGONAL ) {
+            P.orthogonalProjection(-aspect/orthogonalZoom, 
+                                    aspect/orthogonalZoom,
+                                   -1/orthogonalZoom, 1/orthogonalZoom,
+                                   nearPlaneDistance, farPlaneDistance);
+            msg = msg + "  - Projection matrix:" + P;
+        }
+
+        //------------------------------------------------------------
+        return msg;
+    }
+
+    private double fpd()
+    {
+        return (farPlaneDistance - nearPlaneDistance) / nearPlaneDistance;
+    }
+
+    /**
+    Given a point in "clipping coordinates space", this method calculates
+    a six bit opcode, as explained in [FOLE1992].6.5.3, suitable for use in the
+    Cohen-Suterland line clipping algorithm.
+
+    Note that in VSDK, the clipping space correspond to the frustum for
+    the minmax cube from <-1, -1, -1> to <1, 1, 1>.
+
+    WARNING: Currently is only implementing the perspective case!
+    */
+    private int calculateOutcodeBits(Vector3D p)
+    {
+        int bits = 0x00;
+
+        if ( p.z + p.x - 1 > 0 ) bits |= OPCODE_RIGHT;
+        if ( p.z - p.x - 1 > 0 ) bits |= OPCODE_LEFT;
+        if ( p.z + p.y - 1 > 0 ) bits |= OPCODE_UP;
+        if ( p.z - p.y - 1 > 0 ) bits |= OPCODE_DOWN;
+        // Warning: near plane clipping
+        if ( p.z > 0 ) bits |= OPCODE_NEAR;
+        if ( p.z < -fpd() ) bits |= OPCODE_FAR;
+
+        return bits;
+    }
+
+    /**
+    This method implements the Cohen-Sutherland line clipping algorithm with
+    respect to the view volume defined by current camera. Recieves the two
+    line endpoints and return true if any part of this line lies inside the
+    view volume.  In the case the line crosses the view volume, the new
+    resulting endpoints are calculated and returned.
+
+    This algorithm structure follows the one proposed in [FOLE1992].3.12.3,
+    generalizing it to the 3D case, as noted in [FOLE1992].6.5.3.
+
+    The resulting clipped points are in the canonical volume reference, ready
+    for projection.  If 3D clipped points are needed, they must be premultiplied
+    by the inverse of the normalizingTransformation.
+    */
+    public boolean clipLineCohenSutherlandCanonicVolume(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 108 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.